# Import Libraries

In [1]:
# import numpy and pandas libraries
import pandas as pd
import numpy as np
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier 
from matplotlib import pyplot as plt
from sklearn.svm import SVC
from sklearn.metrics import confusion_matrix
from sklearn.model_selection import GridSearchCV, RandomizedSearchCV
np.random.seed(1)
from sklearn.impute import SimpleImputer
# set random seed to ensure that results are repeatable
np.random.seed(1)

In [2]:
from sklearn.metrics import confusion_matrix,accuracy_score,precision_score,recall_score,f1_score

# Loading the Data

In [3]:
X_train = pd.read_csv("X_train_salary_data.csv")
X_test = pd.read_csv("X_test_salary_data.csv")
y_train = pd.read_csv("y_train_salary_data.csv")
y_test = pd.read_csv("y_test_salary_data.csv")

# Model the data

In [4]:
performance = pd.DataFrame({"model": [], "Accuracy": [], "Precision": [], "Recall": [], "F1": []})

In [5]:
X_test.replace([np.inf, -np.inf, np.nan], 0, inplace=True)
y_test.replace([np.inf, -np.inf, np.nan], 0, inplace=True)

# Logistic Regression Model

In [6]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix

In [7]:
log_reg_model = LogisticRegression()
_ = log_reg_model.fit(X_train, np.ravel(y_train))

C:\Users\dkrre\anaconda3\envs\tf\lib\site-packages\sklearn\linear_model\_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [8]:
model_preds = log_reg_model.predict(X_test)
performance = pd.concat([performance, pd.DataFrame({'model':"Logistic Regression Model", 
                                                    'Accuracy': accuracy_score(y_test, model_preds) , 
                                                    'Precision': precision_score(y_test, model_preds, average='weighted') , 
                                                    'Recall': recall_score(y_test, model_preds, average='weighted'), 
                                                    'F1': f1_score(y_test, model_preds, average='weighted') }, index=[0])])
performance

model  Accuracy  Precision    Recall        F1
0  Logistic Regression Model  0.778626   0.778898  0.778626  0.776151

# RandomizedSearchCV Logistic Regression

In [9]:
score_measure = "recall"
LR=LogisticRegression()
kfolds = 5
param_grid = {'C': [0.1, 1, 10,0.001], 
              "solver" : [ 'lbfgs', 'liblinear','saga'],
              "penalty" : ['l1','l2','lasso','elastic']} 
  
grid = RandomizedSearchCV(LR, param_grid, refit = True, verbose = 3)
  
# fitting the model for grid search
grid.fit(X_train, y_train)
print(f"The best {score_measure} score is {grid.best_score_}")
print(f"... with parameters: {grid.best_params_}")

bestRecallTree = grid.best_estimator_

Fitting 5 folds for each of 10 candidates, totalling 50 fits
[CV 1/5] END .C=10, penalty=elastic, solver=lbfgs;, score=nan total time=   0.0s
[CV 2/5] END .C=10, penalty=elastic, solver=lbfgs;, score=nan total time=   0.0s
[CV 3/5] END .C=10, penalty=elastic, solver=lbfgs;, score=nan total time=   0.0s
[CV 4/5] END .C=10, penalty=elastic, solver=lbfgs;, score=nan total time=   0.0s
[CV 5/5] END .C=10, penalty=elastic, solver=lbfgs;, score=nan total time=   0.0s
[CV 1/5] END ...C=1, penalty=elastic, solver=saga;, score=nan total time=   0.0s
[CV 2/5] END ...C=1, penalty=elastic, solver=saga;, score=nan total time=   0.0s
[CV 3/5] END ...C=1, penalty=elastic, solver=saga;, score=nan total time=   0.0s
[CV 4/5] END ...C=1, penalty=elastic, solver=saga;, score=nan total time=   0.0s
[CV 5/5] END ...C=1, penalty=elastic, solver=saga;, score=nan total time=   0.0s
[CV 1/5] END ..C=0.001, penalty=l1, solver=saga;, score=0.327 total time=   0.0s
[CV 2/5] END ..C=0.001, penalty=l1, solver=saga;

C:\Users\dkrre\anaconda3\envs\tf\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\dkrre\anaconda3\envs\tf\lib\site-packages\sklearn\linear_model\_sag.py:350: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
C:\Users\dkrre\anaconda3\envs\tf\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\dkrre\anaconda3\envs\tf\lib\site-packages\sklearn\linear_model\_sag.py:350: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
C:\Users\dkrre\anaconda3\envs\tf\lib\site-packages\sklearn\utils\validation.py:1

[CV 3/5] END ....C=0.1, penalty=l1, solver=saga;, score=0.708 total time=   0.0s
[CV 4/5] END ....C=0.1, penalty=l1, solver=saga;, score=0.604 total time=   0.0s
[CV 5/5] END ....C=0.1, penalty=l1, solver=saga;, score=0.750 total time=   0.0s
[CV 1/5] END ...C=0.1, penalty=l2, solver=lbfgs;, score=0.796 total time=   0.0s
[CV 2/5] END ...C=0.1, penalty=l2, solver=lbfgs;, score=0.571 total time=   0.0s
[CV 3/5] END ...C=0.1, penalty=l2, solver=lbfgs;, score=0.812 total time=   0.0s
[CV 4/5] END ...C=0.1, penalty=l2, solver=lbfgs;, score=0.708 total time=   0.0s


C:\Users\dkrre\anaconda3\envs\tf\lib\site-packages\sklearn\linear_model\_sag.py:350: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
C:\Users\dkrre\anaconda3\envs\tf\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\dkrre\anaconda3\envs\tf\lib\site-packages\sklearn\linear_model\_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
C:\Users\dkrre\ana

[CV 5/5] END ...C=0.1, penalty=l2, solver=lbfgs;, score=0.812 total time=   0.0s
[CV 1/5] END C=10, penalty=lasso, solver=liblinear;, score=nan total time=   0.0s
[CV 2/5] END C=10, penalty=lasso, solver=liblinear;, score=nan total time=   0.0s
[CV 3/5] END C=10, penalty=lasso, solver=liblinear;, score=nan total time=   0.0s
[CV 4/5] END C=10, penalty=lasso, solver=liblinear;, score=nan total time=   0.0s
[CV 5/5] END C=10, penalty=lasso, solver=liblinear;, score=nan total time=   0.0s
[CV 1/5] END .....C=10, penalty=l1, solver=saga;, score=0.735 total time=   0.0s
[CV 2/5] END .....C=10, penalty=l1, solver=saga;, score=0.633 total time=   0.0s
[CV 3/5] END .....C=10, penalty=l1, solver=saga;, score=0.708 total time=   0.0s
[CV 4/5] END .....C=10, penalty=l1, solver=saga;, score=0.604 total time=   0.0s
[CV 5/5] END .....C=10, penalty=l1, solver=saga;, score=0.750 total time=   0.0s
[CV 1/5] END .C=0.001, penalty=lasso, solver=saga;, score=nan total time=   0.0s
[CV 2/5] END .C=0.001, 

C:\Users\dkrre\anaconda3\envs\tf\lib\site-packages\sklearn\linear_model\_sag.py:350: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
C:\Users\dkrre\anaconda3\envs\tf\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\dkrre\anaconda3\envs\tf\lib\site-packages\sklearn\linear_model\_sag.py:350: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
C:\Users\dkrre\anaconda3\envs\tf\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\dkrre\anaconda3\envs\tf\lib\site-packages\sklearn\linear_model\_sag.py:

[CV 3/5] END ....C=10, penalty=l2, solver=lbfgs;, score=0.792 total time=   0.0s
[CV 4/5] END ....C=10, penalty=l2, solver=lbfgs;, score=0.750 total time=   0.0s
[CV 5/5] END ....C=10, penalty=l2, solver=lbfgs;, score=0.812 total time=   0.0s
The best recall score is 0.7606292517006803
... with parameters: {'solver': 'lbfgs', 'penalty': 'l2', 'C': 10}


C:\Users\dkrre\anaconda3\envs\tf\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\dkrre\anaconda3\envs\tf\lib\site-packages\sklearn\linear_model\_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
C:\Users\dkrre\anaconda3\envs\tf\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_sa

In [11]:
model_preds = bestRecallTree.predict(X_test)
performance = pd.concat([performance, pd.DataFrame({'model':"RandomizedSearchCV Logistic Regression", 
                                                    'Accuracy': accuracy_score(y_test, model_preds) , 
                                                    'Precision': precision_score(y_test, model_preds, average='weighted') , 
                                                    'Recall': recall_score(y_test, model_preds, average='weighted'), 
                                                    'F1': f1_score(y_test, model_preds, average='weighted') }, index=[0])])
performance

model  Accuracy  Precision    Recall   
0               Logistic Regression Model  0.778626   0.778898  0.778626  \
0               Logistic Regression Model  0.778626   0.778898  0.778626   
0  RandomizedSearchCV Logistic Regression  0.778626   0.778898  0.778626   

         F1  
0  0.776151  
0  0.776151  
0  0.776151

# GridSearchCV Logistic Regression

In [12]:
score_measure = "recall"
kfolds = 5
param_grid = {'C': [0.1, 1, 10], 
              'solver' : [ 'lbfgs', 'liblinear'],
              'penalty' : ['l1','l2','lasso','elastic']} 
  
grid = GridSearchCV(LogisticRegression(), param_grid, refit = True, verbose = 3)
  
# fitting the model for grid search
grid.fit(X_train, y_train)
print(f"The best {score_measure} score is {grid.best_score_}")
print(f"... with parameters: {grid.best_params_}")

bestRecallTree = grid.best_estimator_

Fitting 5 folds for each of 24 candidates, totalling 120 fits
[CV 1/5] END .....C=0.1, penalty=l1, solver=lbfgs;, score=nan total time=   0.0s
[CV 2/5] END .....C=0.1, penalty=l1, solver=lbfgs;, score=nan total time=   0.0s
[CV 3/5] END .....C=0.1, penalty=l1, solver=lbfgs;, score=nan total time=   0.0s
[CV 4/5] END .....C=0.1, penalty=l1, solver=lbfgs;, score=nan total time=   0.0s
[CV 5/5] END .....C=0.1, penalty=l1, solver=lbfgs;, score=nan total time=   0.0s
[CV 1/5] END C=0.1, penalty=l1, solver=liblinear;, score=0.755 total time=   0.0s
[CV 2/5] END C=0.1, penalty=l1, solver=liblinear;, score=0.653 total time=   0.0s
[CV 3/5] END C=0.1, penalty=l1, solver=liblinear;, score=0.729 total time=   0.0s
[CV 4/5] END C=0.1, penalty=l1, solver=liblinear;, score=0.583 total time=   0.0s
[CV 5/5] END C=0.1, penalty=l1, solver=liblinear;, score=0.771 total time=   0.0s
[CV 1/5] END ...C=0.1, penalty=l2, solver=lbfgs;, score=0.796 total time=   0.0s
[CV 2/5] END ...C=0.1, penalty=l2, solver=

C:\Users\dkrre\anaconda3\envs\tf\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\dkrre\anaconda3\envs\tf\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\dkrre\anaconda3\envs\tf\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\dkrre\anaconda3\envs\tf\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please chan

[CV 4/5] END ...C=0.1, penalty=l2, solver=lbfgs;, score=0.708 total time=   0.0s
[CV 5/5] END ...C=0.1, penalty=l2, solver=lbfgs;, score=0.812 total time=   0.0s
[CV 1/5] END C=0.1, penalty=l2, solver=liblinear;, score=0.776 total time=   0.0s
[CV 2/5] END C=0.1, penalty=l2, solver=liblinear;, score=0.653 total time=   0.0s
[CV 3/5] END C=0.1, penalty=l2, solver=liblinear;, score=0.750 total time=   0.0s
[CV 4/5] END C=0.1, penalty=l2, solver=liblinear;, score=0.625 total time=   0.0s
[CV 5/5] END C=0.1, penalty=l2, solver=liblinear;, score=0.771 total time=   0.0s
[CV 1/5] END ..C=0.1, penalty=lasso, solver=lbfgs;, score=nan total time=   0.0s
[CV 2/5] END ..C=0.1, penalty=lasso, solver=lbfgs;, score=nan total time=   0.0s
[CV 3/5] END ..C=0.1, penalty=lasso, solver=lbfgs;, score=nan total time=   0.0s
[CV 4/5] END ..C=0.1, penalty=lasso, solver=lbfgs;, score=nan total time=   0.0s
[CV 5/5] END ..C=0.1, penalty=lasso, solver=lbfgs;, score=nan total time=   0.0s
[CV 1/5] END C=0.1, pen

C:\Users\dkrre\anaconda3\envs\tf\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\dkrre\anaconda3\envs\tf\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\dkrre\anaconda3\envs\tf\lib\site-packages\sklearn\linear_model\_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linea

[CV 4/5] END .C=1, penalty=l2, solver=liblinear;, score=0.667 total time=   0.0s
[CV 5/5] END .C=1, penalty=l2, solver=liblinear;, score=0.750 total time=   0.0s
[CV 1/5] END ....C=1, penalty=lasso, solver=lbfgs;, score=nan total time=   0.0s
[CV 2/5] END ....C=1, penalty=lasso, solver=lbfgs;, score=nan total time=   0.0s
[CV 3/5] END ....C=1, penalty=lasso, solver=lbfgs;, score=nan total time=   0.0s
[CV 4/5] END ....C=1, penalty=lasso, solver=lbfgs;, score=nan total time=   0.0s
[CV 5/5] END ....C=1, penalty=lasso, solver=lbfgs;, score=nan total time=   0.0s
[CV 1/5] END C=1, penalty=lasso, solver=liblinear;, score=nan total time=   0.0s
[CV 2/5] END C=1, penalty=lasso, solver=liblinear;, score=nan total time=   0.0s
[CV 3/5] END C=1, penalty=lasso, solver=liblinear;, score=nan total time=   0.0s
[CV 4/5] END C=1, penalty=lasso, solver=liblinear;, score=nan total time=   0.0s
[CV 5/5] END C=1, penalty=lasso, solver=liblinear;, score=nan total time=   0.0s
[CV 1/5] END ..C=1, penalty=

C:\Users\dkrre\anaconda3\envs\tf\lib\site-packages\sklearn\svm\_base.py:1244: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(
C:\Users\dkrre\anaconda3\envs\tf\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\dkrre\anaconda3\envs\tf\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\dkrre\anaconda3\envs\tf\lib\site-packages\sklearn\svm\_base.py:1244: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(
C:\Users\dkrre\anaconda3\envs\tf\lib\site-packages\sklearn\utils\validation.py:1143: DataCon

[CV 2/5] END C=10, penalty=l1, solver=liblinear;, score=0.673 total time=   1.9s
[CV 3/5] END C=10, penalty=l1, solver=liblinear;, score=0.792 total time=   0.0s
[CV 4/5] END C=10, penalty=l1, solver=liblinear;, score=0.708 total time=   0.0s
[CV 5/5] END C=10, penalty=l1, solver=liblinear;, score=0.771 total time=   0.0s
[CV 1/5] END ....C=10, penalty=l2, solver=lbfgs;, score=0.837 total time=   0.0s
[CV 2/5] END ....C=10, penalty=l2, solver=lbfgs;, score=0.612 total time=   0.0s


C:\Users\dkrre\anaconda3\envs\tf\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\dkrre\anaconda3\envs\tf\lib\site-packages\sklearn\linear_model\_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
C:\Users\dkrre\anaconda3\envs\tf\lib\site-packages\sklearn\utils\validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_sa

[CV 3/5] END ....C=10, penalty=l2, solver=lbfgs;, score=0.792 total time=   0.0s
[CV 4/5] END ....C=10, penalty=l2, solver=lbfgs;, score=0.750 total time=   0.0s
[CV 5/5] END ....C=10, penalty=l2, solver=lbfgs;, score=0.812 total time=   0.0s
[CV 1/5] END C=10, penalty=l2, solver=liblinear;, score=0.776 total time=   0.0s
[CV 2/5] END C=10, penalty=l2, solver=liblinear;, score=0.612 total time=   0.0s
[CV 3/5] END C=10, penalty=l2, solver=liblinear;, score=0.792 total time=   0.0s
[CV 4/5] END C=10, penalty=l2, solver=liblinear;, score=0.688 total time=   0.0s
[CV 5/5] END C=10, penalty=l2, solver=liblinear;, score=0.750 total time=   0.0s
[CV 1/5] END ...C=10, penalty=lasso, solver=lbfgs;, score=nan total time=   0.0s
[CV 2/5] END ...C=10, penalty=lasso, solver=lbfgs;, score=nan total time=   0.0s
[CV 3/5] END ...C=10, penalty=lasso, solver=lbfgs;, score=nan total time=   0.0s
[CV 4/5] END ...C=10, penalty=lasso, solver=lbfgs;, score=nan total time=   0.0s
[CV 5/5] END ...C=10, penalt

C:\Users\dkrre\anaconda3\envs\tf\lib\site-packages\sklearn\model_selection\_validation.py:378: FitFailedWarning: 
75 fits failed out of a total of 120.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
15 fits failed with the following error:
Traceback (most recent call last):
  File "C:\Users\dkrre\anaconda3\envs\tf\lib\site-packages\sklearn\model_selection\_validation.py", line 686, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "C:\Users\dkrre\anaconda3\envs\tf\lib\site-packages\sklearn\linear_model\_logistic.py", line 1162, in fit
    solver = _check_solver(self.solver, self.penalty, self.dual)
  File "C:\Users\dkrre\anaconda3\envs\tf\lib\site-packages\sklearn\linear_model\_logistic.py", line 54, in _check_s

In [13]:
model_preds = bestRecallTree.predict(X_test)
performance = pd.concat([performance, pd.DataFrame({'model':"GridSearchCV Logistic Regression", 
                                                    'Accuracy': accuracy_score(y_test, model_preds) , 
                                                    'Precision': precision_score(y_test, model_preds, average='weighted') , 
                                                    'Recall': recall_score(y_test, model_preds, average='weighted'), 
                                                    'F1': f1_score(y_test, model_preds, average='weighted') }, index=[0])])
performance

model  Accuracy  Precision    Recall   
0               Logistic Regression Model  0.778626   0.778898  0.778626  \
0               Logistic Regression Model  0.778626   0.778898  0.778626   
0  RandomizedSearchCV Logistic Regression  0.778626   0.778898  0.778626   
0        GridSearchCV Logistic Regression  0.778626   0.778898  0.778626   

         F1  
0  0.776151  
0  0.776151  
0  0.776151  
0  0.776151

# SVM Classification model with Linear Kernel

In [14]:
svm_lin_model = SVC(kernel="linear", probability=True)
_ = svm_lin_model.fit(X_train, np.ravel(y_train))

In [15]:
model_preds = svm_lin_model.predict(X_test)
performance = pd.concat([performance, pd.DataFrame({'model':"SVM Classification model with Linear Kernel", 
                                                    'Accuracy': accuracy_score(y_test, model_preds) , 
                                                    'Precision': precision_score(y_test, model_preds, average='weighted') , 
                                                    'Recall': recall_score(y_test, model_preds, average='weighted'), 
                                                    'F1': f1_score(y_test, model_preds, average='weighted') }, index=[0])])
performance


model  Accuracy  Precision    Recall   
0                    Logistic Regression Model  0.778626   0.778898  0.778626  \
0                    Logistic Regression Model  0.778626   0.778898  0.778626   
0       RandomizedSearchCV Logistic Regression  0.778626   0.778898  0.778626   
0             GridSearchCV Logistic Regression  0.778626   0.778898  0.778626   
0  SVM Classification model with Linear Kernel  0.847328   0.847498  0.847328   

         F1  
0  0.776151  
0  0.776151  
0  0.776151  
0  0.776151  
0  0.846258

# SVM Classification model with rbf Kernel

In [16]:
svm_rbf_model = SVC(kernel="rbf", C=10, gamma='scale', probability=True)
_ = svm_rbf_model.fit(X_train, np.ravel(y_train))

In [17]:
model_preds = svm_rbf_model.predict(X_test)
performance = pd.concat([performance, pd.DataFrame({'model':"SVM Classification model with rbf Kernel", 
                                                    'Accuracy': accuracy_score(y_test, model_preds) , 
                                                    'Precision': precision_score(y_test, model_preds, average='weighted') , 
                                                    'Recall': recall_score(y_test, model_preds, average='weighted'), 
                                                    'F1': f1_score(y_test, model_preds, average='weighted') }, index=[0])])
performance

model  Accuracy  Precision    Recall   
0                    Logistic Regression Model  0.778626   0.778898  0.778626  \
0                    Logistic Regression Model  0.778626   0.778898  0.778626   
0       RandomizedSearchCV Logistic Regression  0.778626   0.778898  0.778626   
0             GridSearchCV Logistic Regression  0.778626   0.778898  0.778626   
0  SVM Classification model with Linear Kernel  0.847328   0.847498  0.847328   
0     SVM Classification model with rbf Kernel  0.702290   0.702525  0.702290   

         F1  
0  0.776151  
0  0.776151  
0  0.776151  
0  0.776151  
0  0.846258  
0  0.700828

## SVM Classification model with Polynomial Kernel

In [18]:
svm_poly_model = SVC(kernel="poly", degree=3, coef0=1, C=10, probability=True)
_ = svm_poly_model.fit(X_train, np.ravel(y_train))

In [19]:
model_preds = svm_poly_model.predict(X_test)
performance = pd.concat([performance, pd.DataFrame({'model':"SVM Classification model with Polynomial Kernel", 
                                                    'Accuracy': accuracy_score(y_test, model_preds) , 
                                                    'Precision': precision_score(y_test, model_preds, average='weighted') , 
                                                    'Recall': recall_score(y_test, model_preds, average='weighted'), 
                                                    'F1': f1_score(y_test, model_preds, average='weighted') }, index=[0])])
performance


model  Accuracy  Precision   
0                        Logistic Regression Model  0.778626   0.778898  \
0                        Logistic Regression Model  0.778626   0.778898   
0           RandomizedSearchCV Logistic Regression  0.778626   0.778898   
0                 GridSearchCV Logistic Regression  0.778626   0.778898   
0      SVM Classification model with Linear Kernel  0.847328   0.847498   
0         SVM Classification model with rbf Kernel  0.702290   0.702525   
0  SVM Classification model with Polynomial Kernel  0.793893   0.797295   

     Recall        F1  
0  0.778626  0.776151  
0  0.778626  0.776151  
0  0.778626  0.776151  
0  0.778626  0.776151  
0  0.847328  0.846258  
0  0.702290  0.700828  
0  0.793893  0.792514

# Decision tree model

In [20]:
classifier = DecisionTreeClassifier()
classifier = classifier.fit(X_train,y_train)

In [21]:
model_preds = classifier.predict(X_test)
performance = pd.concat([performance, pd.DataFrame({'model':"Decision tree model", 
                                                    'Accuracy': accuracy_score(y_test, model_preds) , 
                                                    'Precision': precision_score(y_test, model_preds, average='weighted') , 
                                                    'Recall': recall_score(y_test, model_preds, average='weighted'), 
                                                    'F1': f1_score(y_test, model_preds, average='weighted') }, index=[0])])
performance

model  Accuracy  Precision   
0                        Logistic Regression Model  0.778626   0.778898  \
0                        Logistic Regression Model  0.778626   0.778898   
0           RandomizedSearchCV Logistic Regression  0.778626   0.778898   
0                 GridSearchCV Logistic Regression  0.778626   0.778898   
0      SVM Classification model with Linear Kernel  0.847328   0.847498   
0         SVM Classification model with rbf Kernel  0.702290   0.702525   
0  SVM Classification model with Polynomial Kernel  0.793893   0.797295   
0                              Decision tree model  0.809160   0.811874   

     Recall        F1  
0  0.778626  0.776151  
0  0.778626  0.776151  
0  0.778626  0.776151  
0  0.778626  0.776151  
0  0.847328  0.846258  
0  0.702290  0.700828  
0  0.793893  0.792514  
0  0.809160  0.809896

# Decision tree model using the randomsearch

In [22]:
score_measure = "recall"
kfolds = 5

param_grid = {
    'min_samples_split': np.arange(1,60),  
    'min_samples_leaf': np.arange(1,50),
    'min_impurity_decrease': np.arange(0.0001, 0.01, 0.0005),
    'max_leaf_nodes': np.arange(5, 200), 
    'max_depth': np.arange(1,50), 
    'criterion': ['entropy', 'gini'],
}

dtree = DecisionTreeClassifier()
rand_search = RandomizedSearchCV(estimator = dtree, param_distributions=param_grid, cv=kfolds, n_iter=500,
                           scoring=score_measure, verbose=1, n_jobs=-1, return_train_score=True)

_ = rand_search.fit(X_train, y_train)

print(f"The best {score_measure} score is {rand_search.best_score_}")
print(f"... with parameters: {rand_search.best_params_}")

bestRecallTree = rand_search.best_estimator_

Fitting 5 folds for each of 500 candidates, totalling 2500 fits
The best recall score is nan
... with parameters: {'min_samples_split': 46, 'min_samples_leaf': 10, 'min_impurity_decrease': 0.0086, 'max_leaf_nodes': 68, 'max_depth': 21, 'criterion': 'entropy'}


C:\Users\dkrre\anaconda3\envs\tf\lib\site-packages\sklearn\model_selection\_validation.py:378: FitFailedWarning: 
55 fits failed out of a total of 2500.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
55 fits failed with the following error:
Traceback (most recent call last):
  File "C:\Users\dkrre\anaconda3\envs\tf\lib\site-packages\sklearn\model_selection\_validation.py", line 686, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "C:\Users\dkrre\anaconda3\envs\tf\lib\site-packages\sklearn\tree\_classes.py", line 889, in fit
    super().fit(
  File "C:\Users\dkrre\anaconda3\envs\tf\lib\site-packages\sklearn\tree\_classes.py", line 177, in fit
    self._validate_params()
  File "C:\Users\dkrre\anaconda3\envs\tf\

In [23]:
model_preds = bestRecallTree.predict(X_test)
performance = pd.concat([performance, pd.DataFrame({'model':"Decision tree model using the randomsearch", 
                                                    'Accuracy': accuracy_score(y_test, model_preds) , 
                                                    'Precision': precision_score(y_test, model_preds, average='weighted') , 
                                                    'Recall': recall_score(y_test, model_preds, average='weighted'), 
                                                    'F1': f1_score(y_test, model_preds, average='weighted') }, index=[0])])
performance

model  Accuracy  Precision   
0                        Logistic Regression Model  0.778626   0.778898  \
0                        Logistic Regression Model  0.778626   0.778898   
0           RandomizedSearchCV Logistic Regression  0.778626   0.778898   
0                 GridSearchCV Logistic Regression  0.778626   0.778898   
0      SVM Classification model with Linear Kernel  0.847328   0.847498   
0         SVM Classification model with rbf Kernel  0.702290   0.702525   
0  SVM Classification model with Polynomial Kernel  0.793893   0.797295   
0                              Decision tree model  0.809160   0.811874   
0       Decision tree model using the randomsearch  0.854962   0.860624   

     Recall        F1  
0  0.778626  0.776151  
0  0.778626  0.776151  
0  0.778626  0.776151  
0  0.778626  0.776151  
0  0.847328  0.846258  
0  0.702290  0.700828  
0  0.793893  0.792514  
0  0.809160  0.809896  
0  0.854962  0.855194

# Decision tree model using the Gridsearch

In [24]:
score_measure = "recall"
kfolds = 5

param_grid = {
    'min_samples_split': np.arange(7,10),  
    'min_samples_leaf': np.arange(1,5),
    'min_impurity_decrease': np.arange(0.0048, 0.0054, 0.0001),
    'max_leaf_nodes': np.arange(162,168), 
    'max_depth': np.arange(15,21), 
    'criterion': ['entropy'],
}

dtree = DecisionTreeClassifier()
grid_search = GridSearchCV(estimator = dtree, param_grid=param_grid, cv=kfolds, 
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

_ = grid_search.fit(X_train, y_train)

print(f"The best {score_measure} score is {grid_search.best_score_}")
print(f"... with parameters: {grid_search.best_params_}")

bestRecallTree = grid_search.best_estimator_

Fitting 5 folds for each of 3024 candidates, totalling 15120 fits
The best recall score is nan
... with parameters: {'criterion': 'entropy', 'max_depth': 15, 'max_leaf_nodes': 162, 'min_impurity_decrease': 0.0048, 'min_samples_leaf': 1, 'min_samples_split': 7}


C:\Users\dkrre\anaconda3\envs\tf\lib\site-packages\sklearn\model_selection\_search.py:952: UserWarning: One or more of the test scores are non-finite: [nan nan nan ... nan nan nan]
  warnings.warn(
C:\Users\dkrre\anaconda3\envs\tf\lib\site-packages\sklearn\model_selection\_search.py:952: UserWarning: One or more of the train scores are non-finite: [nan nan nan ... nan nan nan]
  warnings.warn(


In [25]:
model_preds = bestRecallTree.predict(X_test)
performance = pd.concat([performance, pd.DataFrame({'model':"Decision tree model using the Gridsearch", 
                                                    'Accuracy': accuracy_score(y_test, model_preds) , 
                                                    'Precision': precision_score(y_test, model_preds, average='weighted') , 
                                                    'Recall': recall_score(y_test, model_preds, average='weighted'), 
                                                    'F1': f1_score(y_test, model_preds, average='weighted') }, index=[0])])
performance

model  Accuracy  Precision   
0                        Logistic Regression Model  0.778626   0.778898  \
0                        Logistic Regression Model  0.778626   0.778898   
0           RandomizedSearchCV Logistic Regression  0.778626   0.778898   
0                 GridSearchCV Logistic Regression  0.778626   0.778898   
0      SVM Classification model with Linear Kernel  0.847328   0.847498   
0         SVM Classification model with rbf Kernel  0.702290   0.702525   
0  SVM Classification model with Polynomial Kernel  0.793893   0.797295   
0                              Decision tree model  0.809160   0.811874   
0       Decision tree model using the randomsearch  0.854962   0.860624   
0         Decision tree model using the Gridsearch  0.801527   0.806714   

     Recall        F1  
0  0.778626  0.776151  
0  0.778626  0.776151  
0  0.778626  0.776151  
0  0.778626  0.776151  
0  0.847328  0.846258  
0  0.702290  0.700828  
0  0.793893  0.792514  
0  0.809160  0.809896  
0  0.854962  0.855194  
0  0.801527  0.803087

In [26]:
performance.sort_values(by=['Recall'])

model  Accuracy  Precision   
0         SVM Classification model with rbf Kernel  0.702290   0.702525  \
0                        Logistic Regression Model  0.778626   0.778898   
0                        Logistic Regression Model  0.778626   0.778898   
0           RandomizedSearchCV Logistic Regression  0.778626   0.778898   
0                 GridSearchCV Logistic Regression  0.778626   0.778898   
0  SVM Classification model with Polynomial Kernel  0.793893   0.797295   
0         Decision tree model using the Gridsearch  0.801527   0.806714   
0                              Decision tree model  0.809160   0.811874   
0      SVM Classification model with Linear Kernel  0.847328   0.847498   
0       Decision tree model using the randomsearch  0.854962   0.860624   

     Recall        F1  
0  0.702290  0.700828  
0  0.778626  0.776151  
0  0.778626  0.776151  
0  0.778626  0.776151  
0  0.778626  0.776151  
0  0.793893  0.792514  
0  0.801527  0.803087  
0  0.809160  0.809896  
0  0.847328  0.846258  
0  0.854962  0.855194

Using a decision boundary that optimizes the margin between the classes, the widely used classification algorithm SVM with a linear kernel divides data points into distinct classes. The sort of data, the selection of hyperparameters, and the size of the training set are just a few variables that can have an impact on how well an SVM with a linear kernel performs.

When the data is linearly separable—that is, when a hyperplane exists that can completely separate the data points into various classes—SVMs with linear kernels are generally known to perform well. When this is the case, SVMs are able to produce results with high recall and accuracy rates, which makes them a common option for classification tasks.


Decision tree model using the randomsearch is considered to be the best fitting model with a recall value of 85.4%

## Neural networks

In [27]:
from sklearn.neural_network import MLPClassifier

In [36]:
%%time

ann = MLPClassifier(hidden_layer_sizes=(40,20), solver='adam', max_iter=200)
_ = ann.fit(X_train, y_train)

C:\Users\dkrre\anaconda3\envs\tf\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:1098: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


CPU times: total: 0 ns
Wall time: 739 ms


C:\Users\dkrre\anaconda3\envs\tf\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:686: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


In [37]:
%%time
y_pred = ann.predict(X_test)

CPU times: total: 0 ns
Wall time: 0 ns


In [38]:
y_test.value_counts()

Salary
0.0       39
2.0       32
3.0       32
1.0       28
Name: count, dtype: int64

In [39]:
y_pred

array([0, 0, 2, 0, 2, 0, 3, 3, 1, 0, 1, 0, 2, 2, 0, 0, 3, 1, 3, 0, 1, 0,
       2, 3, 3, 3, 0, 3, 0, 0, 3, 3, 0, 3, 0, 3, 3, 0, 2, 3, 1, 0, 2, 3,
       1, 3, 3, 0, 2, 3, 0, 2, 3, 0, 0, 2, 2, 1, 0, 3, 3, 1, 0, 0, 3, 2,
       0, 1, 2, 0, 3, 3, 1, 3, 0, 0, 2, 0, 0, 1, 3, 3, 3, 0, 0, 2, 2, 2,
       2, 3, 3, 3, 1, 3, 1, 0, 2, 2, 1, 2, 3, 3, 1, 1, 1, 2, 2, 0, 0, 0,
       1, 0, 3, 1, 2, 1, 3, 1, 0, 0, 3, 1, 0, 1, 2, 1, 1, 1, 1, 3, 2],
      dtype=int64)

In [42]:
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score, precision_score, recall_score

In [43]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

         0.0       0.85      0.87      0.86        39
         1.0       0.63      0.61      0.62        28
         2.0       0.77      0.62      0.69        32
         3.0       0.79      0.94      0.86        32

    accuracy                           0.77       131
   macro avg       0.76      0.76      0.76       131
weighted avg       0.77      0.77      0.77       131



## Nueral Network model With RandomizedSearchCV

In [49]:
%%time

score_measure = "recall"
kfolds = 5

param_grid = {
    'hidden_layer_sizes': [(40,20),(40,40,40)],
    'activation': ['tanh', 'relu'],
    'solver': ['adam', 'sgd'],
    'alpha': [0, .2, .5, .7, 1],
    'learning_rate': ['constant', 'invscaling', 'adaptive'],
    'learning_rate_init': [0.001, 0.01, 0.1, 0.2, 0.5],
    'max_iter': [5000]
}

ann = MLPClassifier()
grid_search = RandomizedSearchCV(estimator = ann, param_distributions=param_grid, cv=kfolds, n_iter=100,
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

_ = grid_search.fit(X_train, y_train)

bestRecallTree = grid_search.best_estimator_

print(grid_search.best_params_)

Fitting 5 folds for each of 100 candidates, totalling 500 fits


C:\Users\dkrre\anaconda3\envs\tf\lib\site-packages\sklearn\model_selection\_search.py:952: UserWarning: One or more of the test scores are non-finite: [nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
 nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
 nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
 nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
 nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
 nan nan nan nan nan nan nan nan nan nan]
  warnings.warn(
C:\Users\dkrre\anaconda3\envs\tf\lib\site-packages\sklearn\model_selection\_search.py:952: UserWarning: One or more of the train scores are non-finite: [nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
 nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
 nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
 nan nan nan nan nan nan nan nan nan nan nan nan nan n

{'solver': 'sgd', 'max_iter': 5000, 'learning_rate_init': 0.001, 'learning_rate': 'constant', 'hidden_layer_sizes': (40, 40, 40), 'alpha': 0.7, 'activation': 'relu'}
CPU times: total: 1.48 s
Wall time: 43.3 s


In [50]:
%%time
y_pred = bestRecallTree.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

         0.0       0.85      0.87      0.86        39
         1.0       0.47      0.25      0.33        28
         2.0       0.52      0.44      0.47        32
         3.0       0.63      0.97      0.77        32

    accuracy                           0.66       131
   macro avg       0.62      0.63      0.61       131
weighted avg       0.63      0.66      0.63       131

CPU times: total: 15.6 ms
Wall time: 31.2 ms


## Neural Network model With GridSearchCV

In [51]:
%%time

score_measure = "recall"
kfolds = 5

param_grid = {
    'hidden_layer_sizes': [ (40,40,40), (50,40,20), (50,50,50), (80,)],
    'activation': ['relu','tanh'],
    'solver': ['adam'],
    'alpha': [.3, .5, 1],
    'learning_rate': ['adaptive', 'invscaling'],
    'learning_rate_init': [0.005, 0.01, 0.15],
    'max_iter': [5000]
}

ann = MLPClassifier()
grid_search = GridSearchCV(estimator = ann, param_grid=param_grid, cv=kfolds, 
                           scoring=score_measure, verbose=1, n_jobs=-1,  # n_jobs=-1 will utilize all available CPUs 
                           return_train_score=True)

_ = grid_search.fit(X_train, y_train)

bestRecallTree = grid_search.best_estimator_

print(grid_search.best_params_)

Fitting 5 folds for each of 144 candidates, totalling 720 fits


C:\Users\dkrre\anaconda3\envs\tf\lib\site-packages\sklearn\model_selection\_search.py:952: UserWarning: One or more of the test scores are non-finite: [nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
 nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
 nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
 nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
 nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
 nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
 nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
 nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan]
  warnings.warn(
C:\Users\dkrre\anaconda3\envs\tf\lib\site-packages\sklearn\model_selection\_search.py:952: UserWarning: One or more of the train scores are non-finite: [nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan nan
 nan nan nan nan nan n

{'activation': 'relu', 'alpha': 0.3, 'hidden_layer_sizes': (40, 40, 40), 'learning_rate': 'adaptive', 'learning_rate_init': 0.005, 'max_iter': 5000, 'solver': 'adam'}
CPU times: total: 2.2 s
Wall time: 45.9 s


In [52]:
%%time
y_pred = bestRecallTree.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

         0.0       0.90      0.92      0.91        39
         1.0       0.75      0.32      0.45        28
         2.0       0.58      0.69      0.63        32
         3.0       0.76      0.97      0.85        32

    accuracy                           0.75       131
   macro avg       0.75      0.73      0.71       131
weighted avg       0.75      0.75      0.73       131

CPU times: total: 31.2 ms
Wall time: 18.6 ms


In [53]:
X_test.shape

(131, 5)

In [54]:
y_train.value_counts()

Salary
1         73
2         59
0         57
3         53
Name: count, dtype: int64

## Using Keras

## Deep Network

In [118]:
import os
import tensorflow as tf
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3'

In [119]:
import tensorflow as tf
from tensorflow import keras

# fix random seed for reproducibility
np.random.seed(1)
tf.random.set_seed(1)

In [120]:
X_test.shape

(131, 5)

In [121]:
y_test.value_counts()

Salary
0.0       39
2.0       32
3.0       32
1.0       28
Name: count, dtype: int64

In [122]:
%%time

# create model stucture
model = keras.models.Sequential()
model.add(keras.layers.Input(5))
model.add(keras.layers.Dense(50, activation='relu'))
model.add(keras.layers.Dense(50, activation='relu'))
model.add(keras.layers.Dense(50, activation='relu'))
model.add(keras.layers.Dense(4, activation='softmax')) # final layer, 10 categories


# compile
model.compile(loss='sparse_categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

# if you want to overide the defaults for the optimizer....
#adam = keras.optimizers.Adam(learning_rate=0.01)
#model.compile(loss='sparse_categorical_crossentropy', optimizer=adam, metrics=['accuracy'])


CPU times: total: 78.1 ms
Wall time: 94.3 ms


In [124]:
%%time

# fit the model
history = model.fit(X_train, y_train, 
                    validation_data=(X_test, y_test), 
                    epochs=25, batch_size=100)


Epoch 1/25
3/3 [==============================] - 0s 71ms/step - loss: 0.8860 - accuracy: 0.6364 - val_loss: 0.8983 - val_accuracy: 0.6412
Epoch 2/25
3/3 [==============================] - 0s 34ms/step - loss: 0.8689 - accuracy: 0.6405 - val_loss: 0.8847 - val_accuracy: 0.6565
Epoch 3/25
3/3 [==============================] - 0s 39ms/step - loss: 0.8536 - accuracy: 0.6446 - val_loss: 0.8683 - val_accuracy: 0.6489
Epoch 4/25
3/3 [==============================] - 0s 31ms/step - loss: 0.8387 - accuracy: 0.6653 - val_loss: 0.8495 - val_accuracy: 0.6641
Epoch 5/25
3/3 [==============================] - 0s 39ms/step - loss: 0.8207 - accuracy: 0.6570 - val_loss: 0.8272 - val_accuracy: 0.6870
Epoch 6/25
3/3 [==============================] - 0s 32ms/step - loss: 0.8181 - accuracy: 0.6364 - val_loss: 0.8211 - val_accuracy: 0.6565
Epoch 7/25
3/3 [==============================] - 0s 39ms/step - loss: 0.8098 - accuracy: 0.6529 - val_loss: 0.8200 - val_accuracy: 0.6641
Epoch 8/25
3/3 [===========

In [125]:
# evaluate the model

scores = model.evaluate(X_test, y_test, verbose=0)
scores
# In results, first is loss, second is accuracy

[0.6900591850280762, 0.7251908183097839]

In [126]:
# let's format this into a better output...

print("%s: %.2f" % (model.metrics_names[0], scores[0]))
print("%s: %.2f%%" % (model.metrics_names[1], scores[1]*100))

loss: 0.69
accuracy: 72.52%


## Wide and Deep Network

In [128]:
#Define the model: for multi-class

model = keras.models.Sequential()

model.add(keras.layers.Input(shape=5))
model.add(keras.layers.Dense(100, activation='relu'))
model.add(keras.layers.Dense(100, activation='relu'))
model.add(keras.layers.Dense(100, activation='relu'))
model.add(keras.layers.Dense(4, activation='softmax'))

In [129]:
# Compile model

#Optimizer:
adam = keras.optimizers.Adam(learning_rate=0.01)
model.compile(loss='sparse_categorical_crossentropy', optimizer=adam, metrics=['accuracy'])

In [131]:
# Fit the model

history = model.fit(X_train, y_train, 
                    validation_data=(X_test, y_test), 
                    epochs=25, batch_size=100)

Epoch 1/25
3/3 [==============================] - 0s 71ms/step - loss: 0.5369 - accuracy: 0.7314 - val_loss: 0.7051 - val_accuracy: 0.6794
Epoch 2/25
3/3 [==============================] - 0s 32ms/step - loss: 0.5612 - accuracy: 0.7149 - val_loss: 0.5509 - val_accuracy: 0.7405
Epoch 3/25
3/3 [==============================] - 0s 39ms/step - loss: 0.5506 - accuracy: 0.7066 - val_loss: 0.5286 - val_accuracy: 0.7328
Epoch 4/25
3/3 [==============================] - 0s 32ms/step - loss: 0.5280 - accuracy: 0.7521 - val_loss: 0.4992 - val_accuracy: 0.7786
Epoch 5/25
3/3 [==============================] - 0s 39ms/step - loss: 0.4913 - accuracy: 0.7603 - val_loss: 0.4787 - val_accuracy: 0.7710
Epoch 6/25
3/3 [==============================] - 0s 40ms/step - loss: 0.4703 - accuracy: 0.7934 - val_loss: 0.4606 - val_accuracy: 0.7557
Epoch 7/25
3/3 [==============================] - 0s 40ms/step - loss: 0.4903 - accuracy: 0.7645 - val_loss: 0.4723 - val_accuracy: 0.8397
Epoch 8/25
3/3 [===========

In [132]:
# evaluate the model

scores = model.evaluate(X_test, y_test, verbose=0)
scores

# In results, first is loss, second is accuracy

[0.43525972962379456, 0.7938931584358215]

In [133]:
# extract the accuracy from model.evaluate

print("%s: %.2f" % (model.metrics_names[0], scores[0]))
print("%s: %.2f%%" % (model.metrics_names[1], scores[1]*100))

loss: 0.44
accuracy: 79.39%


## Wide and Deep network with RandomGridSearch

In [134]:
%%time

# If you don't have the following installed, from command line '!pip install scikeras'
from scikeras.wrappers import KerasClassifier
from keras.initializers import GlorotNormal

score_measure = "recall"
kfolds = 5

def build_clf(hidden_layer_sizes, dropout):
    ann = tf.keras.models.Sequential()
    ann.add(keras.layers.Input(shape=5)),
    for hidden_layer_size in hidden_layer_sizes:
        model.add(keras.layers.Dense(hidden_layer_size, kernel_initializer= tf.keras.initializers.GlorotNormal(), 
                                     bias_initializer=keras.initializers.RandomNormal(mean=0.0, stddev=0.05, seed=None), activation="relu"))
        model.add(keras.layers.Dropout(dropout))
    ann.add(tf.keras.layers.Dense(4, activation='softmax'))
    ann.compile(loss = 'sparse_categorical_crossentropy', metrics = ['accuracy'])
    return ann

CPU times: total: 0 ns
Wall time: 0 ns


In [135]:
from scikeras.wrappers import KerasClassifier

keras_clf = KerasClassifier(
    model=build_clf,
    hidden_layer_sizes=5,
    dropout = 0.0
)


In [148]:
from tensorflow.keras.callbacks import EarlyStopping
from sklearn.model_selection import RandomizedSearchCV

params = {
    'optimizer__learning_rate': [0.0005, 0.001, 0.005],
    'model__hidden_layer_sizes': [(100,100,100),(100,150,100)],
    'model__dropout': [0, 0.1],
    'batch_size':[20, 60, 100],
    'epochs':[10, 50, 100],
    'optimizer':["adam",'sgd']
}
keras_clf.get_params().keys()


dict_keys(['model', 'build_fn', 'warm_start', 'random_state', 'optimizer', 'loss', 'metrics', 'batch_size', 'validation_batch_size', 'verbose', 'callbacks', 'validation_split', 'shuffle', 'run_eagerly', 'epochs', 'hidden_layer_sizes', 'dropout', 'class_weight'])

In [149]:
rnd_search_cv = RandomizedSearchCV(estimator=keras_clf, param_distributions=params, scoring='accuracy', n_iter=50, cv=5)

import sys
sys.setrecursionlimit(10000) # note: the default is 3000 (python 3.9)

earlystop = EarlyStopping(monitor='val_loss', patience=5, verbose=0, mode='auto')
callback = [earlystop]


In [150]:
rnd_search_cv

RandomizedSearchCV(cv=5,
                   estimator=KerasClassifier(dropout=0.0, hidden_layer_sizes=5, model=<function build_clf at 0x000002308E5F4700>),
                   n_iter=50,
                   param_distributions={'batch_size': [20, 60, 100],
                                        'epochs': [10, 50, 100],
                                        'model__dropout': [0, 0.1],
                                        'model__hidden_layer_sizes': [(100, 100,
                                                                       100),
                                                                      (100, 150,
                                                                       100)],
                                        'optimizer': ['adam', 'sgd'],
                                        'optimizer__learning_rate': [0.0005,
                                                                     0.001,
                                                                     0.005]},
                   scoring='accuracy')

In [ ]:
_ = rnd_search_cv.fit(X_train, y_train, callbacks=callback, verbose=0)

1/1 [==============================] - 0s 53ms/step


1/1 [==============================] - 0s 30ms/step


In [ ]:
rnd_search_cv.best_params_

In [ ]:
best_net = rnd_search_cv.best_estimator_
print(rnd_search_cv.best_params_)

In [ ]:
%%time
y_pred = best_net.predict(X_test)
print(classification_report(y_test, y_pred))

Decision tree model using the randomsearch is considered to be the best fitting model with a accuracy value of 85.4%. 


Accuracies for,

Neural Network 77%

NN with Randomsearch 66%

NN with Gridsearch 75%

Deep network 72.52%

wide and deep network 71%

wide and deep network using randomgridesearch 68%.